data/lib/src/repository/myspace/document_repository_impl.dart

```diff
@@ -54,51 +56,87 @@ class DocumentRepositoryImpl implements DocumentRepository {
+    return _documentDataSources[DataSourceType.LinShareRemote].getAll();
   }
 
   @override
   Future<List<DownloadTaskId>> downloadDocuments(List<DocumentId> documentIds, Token token, Uri baseUrl) {
-    return _documentDataSource.downloadDocuments(documentIds, token, baseUrl);
+    return _documentDataSources[DataSourceType.LinShareRemote].downloadDocuments(documentIds, token, baseUrl);
   }
 
   @override
   Future<List<Share>> share(List<DocumentId> documentIds, List<MailingListId> mailingListIds, List<GenericUser> recipients) {
-    return _documentDataSource.share(documentIds, mailingListIds, recipients);
+    return _documentDataSources[DataSourceType.LinShareRemote].share(documentIds, mailingListIds, recipients);
   }
 
   @override
   Future<String> downloadDocumentIOS(Document document, Token token, Uri baseUrl, CancelToken cancelToken) {
-    return _documentDataSource.downloadDocumentIOS(document, token, baseUrl, cancelToken);
+    return _documentDataSources[DataSourceType.LinShareRemote].downloadDocumentIOS(document, token, baseUrl, cancelToken);
   }
 
   @override
   Future<Document> remove(DocumentId documentId) {
-    return _documentDataSource.remove(documentId);
+    return _documentDataSources[DataSourceType.LinShareRemote].remove(documentId);
   }
 
   @override
   Future<List<Document>> copyToMySpace(CopyRequest copyRequest) {
-    return _documentDataSource.copyToMySpace(copyRequest);
+    return _documentDataSources[DataSourceType.LinShareRemote].copyToMySpace(copyRequest);
   }
 
   @override
   Future<String> downloadPreviewDocument(Document document, DownloadPreviewType downloadPreviewType, Token token, Uri baseUrl, CancelToken cancelToken) {
-    return _documentDataSource.downloadPreviewDocument(document, downloadPreviewType, token, baseUrl, cancelToken);
+    return _documentDataSources[DataSourceType.LinShareRemote].downloadPreviewDocument(document, downloadPreviewType, token, baseUrl, cancelToken);
   }
 
   @override
   Future<Document> rename(DocumentId documentId, RenameDocumentRequest renameDocumentRequest) {
-    return _documentDataSource.rename(documentId, renameDocumentRequest);
+    return _documentDataSources[DataSourceType.LinShareRemote].rename(documentId, renameDocumentRequest);
   }
 
   @override
   Future<DocumentDetails> getDocument(DocumentId documentId) {
-    return _documentDataSource.getDocument(documentId);
+    return _documentDataSources[DataSourceType.LinShareRemote].getDocument(documentId);
   }
 
   @override
   Future<Document> editDescription(DocumentId documentId, EditDescriptionDocumentRequest request) {
-    return _documentDataSource.editDescription(documentId, request);
+    return _documentDataSources[DataSourceType.LinShareRemote].editDescription(documentId, request);
+  }
+
+  @override
+  Future<bool> makeAvailableOffline(Document document, String localPath) {
+    return (_documentDataSources[DataSourceType.Sqlite] as LocalDocumentDataSource)
```

Comment:
that is the problem if the datasource not implement the same contract.

Reply:
I think the DocumentDataSource contains only general datasource related processing. I have inherited LocalDocumentDataSource from DocumentDataSource to make handling the local part easier.